libs/pika/async_mpi/tests/unit/mpi_async_storage.cpp

```diff
@@ -0,0 +1,569 @@
+                             "recv block ", deb::hex<8>(read_slot),
+                             "<- rank", deb::dec<4>(recv_rank),
+                             "tag", deb::hex<4>(tag),
+                             "recv in flight", deb::dec<4>(recvs_in_flight),
+                             "send in flight", deb::dec<4>(sends_in_flight));
+            // clang-format on
+            void* buffer_to_recv = &local_recv_storage[memory_offset_recv];
+            auto rsnd1 = mpi::transform_mpi(
+                ex::just(buffer_to_recv, options.transfer_size_B,
+                    MPI_UNSIGNED_CHAR, recv_rank, tag, MPI_COMM_WORLD),
+                MPI_Irecv);
+            auto rsnd2 = rsnd1 | ex::then([&](int result) {
+                --recvs_in_flight;
+                nws_deb<5>.debug(deb::str<>("recv complete"), "recv in flight",
+                    recvs_in_flight, "send in flight", sends_in_flight);
+                return result;
+            });
+            ex::start_detached(rsnd2);
+
+            // we are about to send, so increment our counter
+            ++sends_in_flight;
+            // clang-format off
+            nws_deb<5>.debug(deb::str<>("posting send"),
+                             "Rank ", deb::dec<3>(rank),
+                             "send block ", deb::hex<8>(write_slot),
+                             "-> rank", deb::dec<4>(send_rank),
+                             "tag", deb::hex<4>(tag),
+                             "recv in flight", deb::dec<4>(recvs_in_flight),
+                             "send in flight", deb::dec<4>(sends_in_flight));
+            // clang-format on
+            void* buffer_to_send = &local_send_storage[memory_offset_send];
+            auto ssnd1 = mpi::transform_mpi(
+                ex::just(buffer_to_send, options.transfer_size_B,
+                    MPI_UNSIGNED_CHAR, send_rank, tag, MPI_COMM_WORLD),
+                MPI_Isend);
+            auto ssnd2 = ssnd1 | ex::then([&](int result) {
+                --sends_in_flight;
+                nws_deb<5>.debug(deb::str<>("send complete"), "recv in flight",
+                    recvs_in_flight, "send in flight", sends_in_flight);
+                return result;
+            });
+            ex::start_detached(ssnd2);
+        }
+        messages_sent++;
+        //
+        if (RANK_OUTPUT && debug_timer.trigger())
+        {
+            std::cout << ((debug_tick++ % 10 == 9) ? "x" : ".") << std::flush;
+        }
+
+        bool time_up = (exec_timer.elapsed() >= options.num_seconds);
+
+        //
+        if (time_up && !count_complete)
+        {
+            for (uint32_t i = 0; i < nranks; ++i)
+            {
+                if (i != rank)
+                {
+                    MPI_Irecv(&counts[i], 1, MPI_INT32_T, i, 0, MPI_COMM_WORLD,
+                        &recvsr[i]);
+                    MPI_Isend(&messages_sent, 1, MPI_INT32_T, i, 0,
+                        MPI_COMM_WORLD, &sendsr[i]);
+                }
```

Comment:
These messages are "outside of the test" in a sense. The test lets the ranks run for N seconds, sending and receiving as much as they can, but not synchronized. At the end of N seconds, each rank sends the number of messages it posted and the others have to then post any missing receives. If this used the sender/receiver then it would also go through throttling and could cause a deadlock - so it has to be "out of bound" messagiing to ensure it always goes through.